### ISSUE TITLE
Support Typescript incremental build

### ISSUE THREAD

**Issue by glebmachine** [angular/angular#42011]
# 🚀 Feature request
```
- [ ] new
- [x] build
- [x] serve
- [x] test
- [x] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Description
Support of Typescript incremental build https://devblogs.microsoft.com/typescript/announcing-typescript-3-4-rc/

**Comment by glebmachine** [comment]
@alan-agius4 awesome news! 
Hope you will implement support soon

let me know if I can help you, that is very important because of slow rebuilds on large projects

**Comment by clydin** [comment]
Rebuilds for a running instance (`ng build --watch`/`ng serve`) are already incremental.  The support referenced in this issue would only affect _warm_ build scenarios (i.e., repeat execution of `ng build`/`ng serve`).

**Comment by glebmachine** [comment]
Ok, what we do then with 40-60s rebuild time on ‘ng serve’?

angular ignores microfrontends for several years, bazel initiative was closed before it goes to production.

Angular will die if we will keep ignored that.
That’s a real problem!

we don’t need angular designer and features that cover corner cases

But, what is community real needs:
- fast build rebuild
- Ssr support out of the box (like next/next do)
- Reduce bundle size (big angular drama)

For now, angular only fit to cms control panels, but lost competition to all public projects with lighthouse metrics, search engine optimizations and other common needs

**Comment by clydin** [comment]
Please try to keep the issue on the original topic.

If you have cases of 40-60 second application rebuilds, information regarding the project would greatly be appreciated by the team.  If possible please open a separate issue detailing the performance situation that is being encountered.  Even without access to the application code itself, the team has had success troubleshooting performance concerns when performance profiles have been provided.  If needed, a team member can also provide assistance with capturing the performance profiles for analysis once the issue has been opened.

**Comment by mlc-mlapis** [comment]
The most important time is the incremental rebuild during the development process that is repeated hundred times.

**Comment by alan-agius4** [comment]
> The most important time is the incremental rebuild during the development process that is repeated hundred times.

That’s not really what this issue is about. Incremental TypeScript API only improves subsequent cold builds as retrieves the old program data from disk.

**Comment by mlc-mlapis** [comment]
Yep, that's why I wanted to stress where the importance lies because `Incremental TypeScript API` can help at some point, but it's not crucial.

**Comment by simeyla** [comment]
> That’s not really what this issue is about. Incremental TypeScript API only improves subsequent cold builds as retrieves the old program data from disk.

Isn't that a little naïve? When you change something after running `ng serve` it does an incremental build, and sometimes that incremental build takes 1 seconds and sometimes it takes 10. It depends what you change and what kind of ripple through effect that has through the rest of your code. It's not like you change 2 single lines of code in your source and then the angular compiler gives you 2 lines of updated .js files to patch.

When you change something that a lot of other pieces of your application depend upon there'll be a lot of unnecessary recompilation throughout the 'recompilation tree'. All for the incremental change that looked trivial. Surely some of that compilation would benefit from the Typescript incremental cache. 

Is this issue basically dead? It's 3 years old, and while I periodically find it and read through the updates it looks like it's effectively been abandoned. I just don't buy the excuse that 'ng serve already does incremental so it wouldn't have much effect`. I realize these things are always complex, but saving 1 second that adds up. Also I would **love** for a cold build to be faster!

Q: When am I most likely to get distracted and pick up my phone? 
A: When waiting for an incremental build.

**Comment by maxisam** [comment]
@simeyla if you want, you can try nx. If a project is big enough, I think NX is the way to go. Angular CI is good until the size of the project is too big

**Comment by alxhub** [comment]
Next step: followup with our tooling team and determine the potential benefits of this (vs. the rather large implementation cost)

**Comment by varthc** [comment]
Hi! Is there any update on this? Is it planned to include it? Thanks!

**Comment by a-boertien** [comment]
Is this now supported for Angular 18?

I received the following error on upgrading from 17-> 18 and the new build system (esbuild)

>  Option 'tsBuildInfoFile' cannot be specified without specifying option 'incremental' or option 'composite'

I don't have the tsBuildInfoFile option set in my tsconfig (and `tsc --showConfig` also shows this),
but setting `incremental: true` seems to fix it ?

**Comment by JoostK** [comment]
> Is this now supported for Angular 18?
> 
> I received the following error on `ng build` when upgrading from 17-> 18 and the new build system (esbuild)
> 
> > Option 'tsBuildInfoFile' cannot be specified without specifying option 'incremental' or option 'composite'
> 
> I don't have the tsBuildInfoFile option set in my tsconfig (and `tsc --showConfig` also shows this), but setting `incremental: true` seems to fix it ?

The CLI does indeed use incremental TS builds, if caching is enabled. OTOH, `ngc` (the raw variant of `tsc` that includes the Angular compiler) does not.

The CLI does set `incremental` using a nullish-assignment, so if you used to have it set to `false` before then that would explain that error.

https://github.com/angular/angular-cli/blob/d257388fee658e06fd4e09e369f9624db22e72c1/packages/angular/build/src/tools/esbuild/angular/compiler-plugin.ts#L542-L547
